osgp/platform/osgp-adapter-domain-core/src/main/java/org/opensmartgridplatform/adapter/domain/core/application/services/FirmwareManagementService.java

```diff
@@ -337,24 +438,41 @@ public void tryToAddFirmwareVersionToHistory(final String deviceIdentification,
         boolean recordAdded = false;
 
         for (final FirmwareFile file : firmwareFiles) {
-            final Map<FirmwareModule, String> moduleVersions = file.getModuleVersions();
-            if (moduleVersions.containsKey(module) && moduleVersions.get(module).equals(firmwareVersion.getVersion())) {
-
-                // file found, insert a record into the history
-                final DeviceFirmwareFile deviceFirmwareFile = new DeviceFirmwareFile(device, file, new Date(),
-                        INSTALLER);
-                this.deviceFirmwareFileRepository.save(deviceFirmwareFile);
-                LOGGER.info("Firmware version {} added to device {}", firmwareVersion.getVersion(),
-                        deviceIdentification);
-
-                // we only want to add one record in history
-                recordAdded = true;
-                break;
+            int numberOfModulesFound = 0;
+            for (FirmwareVersion firmwareNotCurrent : firmwareVersionsNotCurrent) {
+                final FirmwareModule module = createFirmwareModule(firmwareNotCurrent);
+                final Map<FirmwareModule, String> moduleVersions = file.getModuleVersions();
+                if (moduleVersions.containsKey(module)
+                        && moduleVersions.get(module).equals(firmwareNotCurrent.getVersion())) {
+                    // module found in this file
+                    numberOfModulesFound++;
+                    LOGGER.info(
+                            "{} of {} modules found, Firmware version {} for moduleType {} found for to device {} in File {}, ",
+                            numberOfModulesFound, firmwareVersionsNotCurrent.size(), firmwareNotCurrent.getVersion(),
+                            firmwareNotCurrent.getFirmwareModuleType().toString(), deviceIdentification,
+                            file.getFilename());
+
+                    // check if all different modules are in this file
+                    if (numberOfModulesFound == firmwareVersionsNotCurrent.size()) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3AS134)